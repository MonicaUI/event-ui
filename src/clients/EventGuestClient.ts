import { RequestBuilder, fetchAndRetryIfServerBusy } from '../utils/wrappedFetchAndRetryIfServerTooBusy';
import { SESSION_HEADER_EVENT_ID } from '@cvent/event-ui-networking';
import { ServiceError } from '@cvent/event-ui-networking';
import { convertResponse } from './RegCartClient';
import Logger from '@cvent/nucleus-logging';

const LOG = new Logger('EventGuestClient');

async function fetchOrThrow(request, errorMessage) {
  const response = await fetchAndRetryIfServerBusy(request);
  if (!response.ok) {
    throw await ServiceError.create(errorMessage, response, request);
  }
  if (response.status === 204) {
    return;
  }
  return await response.json();
}

/**
 * The client to manage a guest for the event.
 */
export default class EventGuestClient {
  baseUrl: $TSFixMe;
  environment: $TSFixMe;
  eventId: $TSFixMe;
  constructor(baseUrl?: $TSFixMe, eventId?: $TSFixMe, environment?: $TSFixMe) {
    this.baseUrl = baseUrl;
    this.eventId = eventId;
    this.environment = environment;
  }
  _getRequestBuilder(authToken: $TSFixMe, urlPath = ''): $TSFixMe {
    const builder = new RequestBuilder({ url: `${this.baseUrl}${urlPath}` })
      .header(SESSION_HEADER_EVENT_ID, this.eventId)
      .auth(authToken)
      .withCookies();
    return this.environment ? builder.query('environment', this.environment) : builder;
  }
  /**
   * Returns the baseUrl for the event guest site
   */
  getBaseUrl = (): $TSFixMe => this.baseUrl;

  /**
   * Logs out the user dissasociating the user from any saved information in their session.
   */
  async logout(authToken: $TSFixMe, eventId: $TSFixMe, userType: $TSFixMe): Promise<$TSFixMe> {
    const request = this._getRequestBuilder(authToken, `event/${eventId}`).query('userType', userType).delete().build();
    return await fetchOrThrow(request, 'Failed to logout user.');
  }

  /**
   * Logs user out and aborts an ongoing INPROGRESS regCart using Navigator.sendBeacon API
   */
  async abortRegCartAndLogout(eventId: $TSFixMe, userType: $TSFixMe): Promise<$TSFixMe> {
    const request = this._getRequestBuilder(null, 'reg/regAbort')
      .query('environment', this.environment)
      .query('eventId', eventId)
      .query('userType', userType)
      .build();

    const response = navigator.sendBeacon(request.url);
    if (!response) {
      LOG.info('Error logging out and aborting reg cart ', request.url);
    }
  }

  /**
   * Issues request to rescind earlier abort registration and logout request
   */
  async rescindAbortRegCartAndLogoutRequest(
    authToken: $TSFixMe,
    eventId: $TSFixMe,
    userType: $TSFixMe
  ): Promise<$TSFixMe> {
    LOG.debug('rescindAbortRegCartAndLogoutRequest');
    const request = this._getRequestBuilder(authToken, 'reg/rescindRegAbort')
      .query('environment', this.environment)
      .query('eventId', eventId)
      .query('userType', userType)
      .post()
      .build();

    return await fetchOrThrow(request, 'Failed to rescind registration abort.');
  }

  /**
   * Opt out or opt in email communications of all events under an give account.
   * @param authToken
   * @param {string} eventId - an event ID from the desired opt out/in account
   * @param {string} inviteeId - invitee ID to identify the user from the event
   * @param {string} guestId - guest ID to identify the guest from the event
   * @param {string} contactId - contact ID to identify the contact from the event
   */
  async getOptOutStatus(
    authToken: $TSFixMe,
    eventId: $TSFixMe,
    inviteeId: $TSFixMe,
    guestId: $TSFixMe,
    contactId: $TSFixMe
  ): Promise<$TSFixMe> {
    const request = this._getRequestBuilder(authToken, `event/${eventId}/opt-in-out`)
      .query('environment', this.environment)
      .query('inviteeId', inviteeId)
      .query('guestId', guestId)
      .query('contactId', contactId)
      .get();
    return await fetchOrThrow(request.build(), 'Failed to retrieve user opt-out status.');
  }

  /**
   * Opt out or opt in email communications of all events under an give account.
   * @param authToken
   * @param {string} eventId - an event ID from the desired opt out/in account
   * @param {string} inviteeId - invitee ID to identify the user from the event
   * @param {string} guestId - guest ID to identify the guest from the event
   * @param {string} contactId - contact ID to identify the contact from the event
   * @param {boolean} optOutStatus - desired optOut status. true - opted out.
   */
  async optOut(
    authToken: $TSFixMe,
    eventId: $TSFixMe,
    inviteeId: $TSFixMe,
    guestId: $TSFixMe,
    contactId: $TSFixMe,
    optOutStatus: $TSFixMe
  ): Promise<$TSFixMe> {
    const request = this._getRequestBuilder(authToken, `event/${eventId}/opt-in-out`)
      .query('environment', this.environment)
      .query('inviteeId', inviteeId)
      .query('guestId', guestId)
      .query('contactId', contactId)
      .query('optOut', optOutStatus)
      .put();
    return await fetchOrThrow(request.build(), 'Failed to opt-out user.');
  }

  /**
   * Get registration content (pages, layout items, etc.)
   */
  async getRegistrationContent(
    eventId: $TSFixMe,
    snapshotVersion: $TSFixMe,
    pageVariety: $TSFixMe,
    regPathId = null,
    registrationTypeId = null
  ): Promise<$TSFixMe> {
    const regPathSegment = regPathId ? `registrationPath/${regPathId}` : 'defaultRegistrationPath';
    const request = this._getRequestBuilder(null, `websiteContent/${eventId}/${regPathSegment}/${pageVariety}`)
      .get()
      .query('snapshotVersion', snapshotVersion)
      .query('registrationTypeId', registrationTypeId)
      .build();
    const response = await fetchAndRetryIfServerBusy(request);
    if (!response.ok) {
      throw await ServiceError.create(`Event snapshot load failed eventId: ${String(eventId)}`, response, request);
    }
    return await response.json();
  }

  /**
   * Get website page content (pages, layout items, etc.)
   */
  async getWebsiteContent(eventId: $TSFixMe, snapshotVersion: $TSFixMe, registrationTypeId = null): Promise<$TSFixMe> {
    const request = this._getRequestBuilder(null, `websiteContent/${eventId}/website`)
      .get()
      .query('snapshotVersion', snapshotVersion)
      .query('registrationTypeId', registrationTypeId)
      .build();
    const response = await fetchAndRetryIfServerBusy(request);
    if (!response.ok) {
      throw await ServiceError.create(`Event snapshot load failed eventId: ${String(eventId)}`, response, request);
    }
    return await response.json();
  }

  /**
   * Get site editor core content
   */
  async getSiteEditorCore(
    eventId: $TSFixMe,
    snapshotVersion: $TSFixMe,
    registrationPathId?: $TSFixMe
  ): Promise<$TSFixMe> {
    const request = this._getRequestBuilder(null, `websiteContent/${eventId}/siteEditorCore`)
      .get()
      .query('snapshotVersion', snapshotVersion)
      .query('registrationPathId', registrationPathId)
      .build();
    const response = await fetchAndRetryIfServerBusy(request);
    if (!response.ok) {
      throw await ServiceError.create(
        `Event siteEditorCore load failed eventId: ${String(eventId)}`,
        response,
        request
      );
    }
    return await response.json();
  }

  /**
   * Get documents for a given speaker
   */
  async getSpeakerDocuments(
    authToken: string,
    eventId: string,
    snapshotVersion: string,
    speakerId: string,
    flexBearerAuthRemoval = false
  ): Promise<Record<string, unknown> | ServiceError> {
    let request = this._getRequestBuilder(authToken, `event/speaker/${eventId}/speakerDocuments`);

    if (flexBearerAuthRemoval) {
      request = this._getRequestBuilder(null, `event/speaker/${eventId}/speakerDocuments`);
    }

    request = request.get().query('snapshotVersion', snapshotVersion).query('speakerId', speakerId).build();

    const response = await fetchAndRetryIfServerBusy(request);
    if (!response.ok) {
      throw await ServiceError.create(
        `Failed to retrieve speaker documents for eventId: ${String(eventId)} and speakerId: ${String(speakerId)}`,
        response,
        request
      );
    }
    return await response.json();
  }

  async identifyByContactId(authToken: $TSFixMe, eventId: $TSFixMe, contactId?: $TSFixMe): Promise<$TSFixMe> {
    const request = this._getRequestBuilder(authToken, 'reg/identifyByContactId')
      .query('environment', this.environment)
      .post()
      .json({ eventId, contactId })
      .build();
    const response = await fetchAndRetryIfServerBusy(request);
    if (!response.ok) {
      const error = await ServiceError.create('identifyByContactId failed', response, request);
      LOG.error('Error identifying by contact id', request.url, error.responseStatus, error.responseBody);
      throw error;
    }
    return convertResponse(await response.json());
  }

  /**
   * Get related contacts list containing firstName, lastName, emailAddress and relatedContactStub for a contact.
   * @param authToken
   * @param {string} eventId - event ID of the current registration
   * @param {string} regCartId - regCart ID of the current registration
   * @param {string} contactId - contact ID of the invitee for whom guests will be added from related contacts
   * @param {int} limit - number of related contacts we want to fetch and show on the guests popup
   * @param {null} searchString - search criteria to filter the related contacts to show on the guests popup
   */
  async getRelatedContacts(
    authToken: $TSFixMe,
    eventId: $TSFixMe,
    regCartId: $TSFixMe,
    contactId: $TSFixMe,
    limit: $TSFixMe,
    searchString?: $TSFixMe
  ): Promise<$TSFixMe> {
    const request = this._getRequestBuilder(authToken, `contacts/${eventId}/relatedContacts`)
      .query('regCartId', regCartId)
      .query('contactId', contactId)
      .query('limit', limit)
      .query('searchString', searchString)
      .query('environment', this.environment)
      .get()
      .build();
    const response = await fetchAndRetryIfServerBusy(request);
    if (!response.ok) {
      const error = await ServiceError.create('getRelatedContacts failed', response, request);
      LOG.info('Error getting related contacts', request.url, error.responseStatus, error.responseBody);
      throw error;
    }
    return await response.json();
  }

  /**
   * Publishes an attendee activity fact. Purposefully not throwing the error should one occur. Since it's analytic
   * tracking, we want to log the error and let the application continue.
   */
  async publishAttendeeActivityFact(authToken: $TSFixMe, fact: $TSFixMe): Promise<$TSFixMe> {
    const body = {
      activities: [fact]
    };
    const request = this._getRequestBuilder(authToken, 'attendeeactivities/v1/attendeeactivities/publish')
      .query('environment', this.environment)
      .post()
      .json(body)
      .build();
    const response = await fetchAndRetryIfServerBusy(request);

    if (!response.ok) {
      const error = await ServiceError.create('attendeeactivities/publish failed', response, request);
      LOG.info('Error publishing fact', request.url, error.responseStatus, error.responseBody);
    }
  }

  async getEventWebAppStatus(authToken: $TSFixMe): Promise<$TSFixMe> {
    const request = this._getRequestBuilder(authToken, 'event-hub/published')
      .query('environment', this.environment)
      .query('eventId', this.eventId)
      .build();
    const response = await fetchAndRetryIfServerBusy(request);
    if (!response.ok) {
      const error = await ServiceError.create('getEventWebAppStatus failed', response, request);
      LOG.info('Error getting event webapp status', request.url, error.responseStatus, error.responseBody);
      throw error;
    }
    return await response.json();
  }
}
