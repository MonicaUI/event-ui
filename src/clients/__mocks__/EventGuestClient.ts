export default class MockEventGuestClient {}

Object.assign(MockEventGuestClient.prototype, {
  logout: jest.fn(() => Promise.resolve()),
  getOptOutStatus: jest.fn(() => Promise.resolve()),
  optOut: jest.fn(() => Promise.resolve()),
  getRegistrationContent: jest.fn(() =>
    Promise.resolve({
      registrationPathId: 'regPathId',
      pageVariety: 'REGISTRATION',
      pageIds: [],
      pages: {},
      layoutItems: {},
      registrationPath: {
        id: 'regPathId',
        registrationPageFields: [],
        modification: {}
      }
    })
  ),
  getWebsiteContent: jest.fn(() => Promise.resolve({}))
});
