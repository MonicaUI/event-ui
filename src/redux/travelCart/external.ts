import Logger from '@cvent/nucleus-logging';
import { getIn } from 'icepick';
import {
  travelCartUpdated,
  clearUserSessionData,
  updateUnsavedCreditCard,
  removeUnSavedTravelAnswers
} from './actions';
import { SET_SELECTED_AIR_REQUEST_ID, SET_SELECTED_AIR_ACTUAL_ID, SET_SELECTED_GROUP_FLIGHT_ID } from './actionTypes';
import { updateTravelCartAndUpdateInState, persistTravelCartAndUpdateState, updateTravelCreditCard } from './internal';
import { getRegCart } from '../selectors/shared';
import { loadAirportsForTravelCart } from '../airports';
import { getPrimaryRegistrationId, getAttendeeId, getEventRegistrations } from '../registrationForm/regCart/selectors';
import { loadAirportsForEventAndTravelCart } from '../airports';
import { updateRegistrantInformationInTravelBookings, hasAnyTravelItem } from '../../utils/travelUtils';
import { getTravelCartSnapshotVersion } from './selectors';
import { loadTravelSnapshotVersion } from 'event-widgets/redux/modules/eventTravel';
import { updateTravelAnswers } from './travelAnswers';

const LOG = new Logger('redux/travelCart/external');

/**
 * Restores the travel cart from the backend into the state
 * @param {*} travelCartId The travel cart id corresponding to which the travel cart should be loaded
 */
export function restoreTravelCartIntoState(travelCartId: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { travelApiClient }
    } = getState();
    try {
      const travelCart = await travelApiClient.getTravelCart(travelCartId);
      await dispatch(travelCartUpdated(travelCart));
      await dispatch(clearUserSessionData());
      await dispatch(loadTravelSnapshotVersion(getTravelCartSnapshotVersion(getState())));
      await dispatch(loadAirportsForEventAndTravelCart());
      LOG.debug(`restore travel cart with id ${travelCartId} was successful`);
    } catch (error) {
      LOG.error(`restore travel cart with id ${travelCartId} failed`, error);
      throw error;
    }
  };
}

/**
 * Persists the current travel cart in the state to backend
 * Used for making sure that the travel cart stays alive during the registration
 */
function persistTravelCart(travelForm, saveTravelCreditCard) {
  return async (dispatch, getState) => {
    const { cart, isCartCreated } = travelForm;
    // persist travel credit card
    let updatedCart = await dispatch(updateTravelCreditCard(cart, saveTravelCreditCard));
    // persist travelAnswers
    updatedCart = await dispatch(updateTravelAnswers(updatedCart));
    await persistTravelCartAndUpdateState(updatedCart, isCartCreated, dispatch, getState());
    await dispatch(updateUnsavedCreditCard(null)); // Clear credit card from userSession object
    await dispatch(removeUnSavedTravelAnswers()); // Clear travel answers from userSession object
  };
}

/**
 * Update reg type and admission item ids in travel cart corresponding to the values in reg cart
 * if anything has changed
 */
export function updateRegTypeAndAdmissionItemIdsInTravelBookings() {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const {
      clients: { travelApiClient },
      travelCart: { cart, isCartCreated }
    } = state;
    if (!isCartCreated) {
      return;
    }
    const regCart = getRegCart(state);
    const modifiedResult = updateRegistrantInformationInTravelBookings(cart.bookings, regCart);
    if (modifiedResult.hasAnyBookingChanged) {
      await updateTravelCartAndUpdateInState(
        travelApiClient,
        {
          ...cart,
          bookings: modifiedResult.bookings
        },
        dispatch
      );
    }
  };
}

/**
 * Restores the transient travel cart corresponding to the regcart in state, if applicable
 */
export function restoreTransientTravelCartIntoState(forceLoadTransientCart?: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const {
      clients: { travelApiClient },
      event,
      defaultUserSession: { isPlanner, isTestMode }
    } = state;
    const regCart = getRegCart(state);

    const travelConfigs = getIn(event, ['eventFeatureSetup', 'travelAndHousing']);
    if (
      ((regCart && !regCart.regCartId) || forceLoadTransientCart) &&
      travelConfigs &&
      (travelConfigs.hotelAccomodations || getIn(travelConfigs, ['airTravel', 'enabled']) || travelConfigs.carRental)
    ) {
      let bookingIdOverrides = {};
      const eventRegistrations = getEventRegistrations(regCart);
      Object.keys(eventRegistrations).forEach(
        eventRegId =>
          (bookingIdOverrides = {
            ...bookingIdOverrides,
            [getAttendeeId(regCart, eventRegId)]: eventRegId
          })
      );
      try {
        const response = await travelApiClient.getTransientTravelCart(
          getAttendeeId(regCart, getPrimaryRegistrationId(regCart)),
          true,
          { isForPlanner: isPlanner, isForTestMode: isTestMode, travelBookingIdOverrides: bookingIdOverrides }
        );
        // if response is successful set travelCart into state
        if (response.travelCart) {
          await dispatch(travelCartUpdated(response.travelCart));
          dispatch(loadAirportsForTravelCart());
        }
      } catch (error) {
        if (!(error.responseStatus && error.responseStatus === 404)) {
          throw error;
        }
      }
    }
  };
}

const requireSaving = (regCart, travelForm) => {
  return hasAnyTravelItem(travelForm) && !(regCart.regCancel && travelForm.cart.status === 'TRANSIENT');
};

/**
 * saves the current travel registration form to backend if needed
 * @param saveTravelCreditCard
 * @returns {Function}
 */
export function saveTravelRegistration(saveTravelCreditCard: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      travelCart: travelForm,
      registrationForm: { regCart }
    } = getState();
    if (requireSaving(regCart, travelForm)) {
      await dispatch(persistTravelCart(travelForm, saveTravelCreditCard));
    }
  };
}

/**
 * Retrieves a TravelCart from service when restoring the registration form
 * @param regCart
 * @returns {Function}
 */
export const restoreTravelRegistration = (regCart: $TSFixMe) => {
  return async (dispatch: $TSFixMe): Promise<$TSFixMe> => {
    // when invitee starts reg cancel, we load the transient cart, so can again load transient cart
    if (regCart.regCancel) {
      await dispatch(restoreTransientTravelCartIntoState(true));
    } else if (regCart.hasTravel) {
      await dispatch(restoreTravelCartIntoState(regCart.regCartId));
    }
  };
};

/**
 * start cancellation flow for travel
 * @returns {Function}
 */
export const startTravelCancellation = () => {
  return async (dispatch: $TSFixMe): Promise<$TSFixMe> => {
    /*
     * The actual cancel travel cart has requestStatus props for all bookings set to CANCEL, from that
     * we won't be able to tell whether the bookings were already cancelled or are coming for getting cancelled.
     * Therefore we're using the transient travel cart, where those props would be accurate for our purpose.
     * In case later we need any data specific to travel cancellation, we'd require loading the actual cancel cart then.
     */
    dispatch(restoreTransientTravelCartIntoState(true));
  };
};

/**
 * To add value to selectedGroupFlightIds in state
 */
export function setSelectedGroupFlightId(selectedGroupFlightId: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedGroupFlightIds = [...getState().travelCart.userSession.groupFlights.selectedGroupFlightIds];
    if (!selectedGroupFlightIds.includes(selectedGroupFlightId)) {
      selectedGroupFlightIds.push(selectedGroupFlightId);
      dispatch({
        type: SET_SELECTED_GROUP_FLIGHT_ID,
        payload: {
          selectedGroupFlightIds
        }
      });
    }
  };
}

/**
 * To remove value from selectedGroupFlightIds in state
 * @param {Arrar<string>} idsToRemove
 */
export function removeSelectedGroupFlightIds(idsToRemove = []) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedGroupFlightIds = getState().travelCart.userSession.groupFlights.selectedGroupFlightIds || [];

    if (selectedGroupFlightIds.length > 0 && idsToRemove.length > 0) {
      const filteredIds = selectedGroupFlightIds.filter(existingId => {
        if (!idsToRemove.includes(existingId)) {
          return existingId;
        }
      });
      dispatch({
        type: SET_SELECTED_GROUP_FLIGHT_ID,
        payload: {
          selectedGroupFlightIds: filteredIds
        }
      });
    }
  };
}

/**
 * To add value to selectedAirRequestIds in state
 */
export function setSelectedAirRequestId(selectedAirRequestId: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedAirRequestIds = [...getState().travelCart.userSession.airRequest.selectedAirRequestIds];
    if (!selectedAirRequestIds.includes(selectedAirRequestId)) {
      selectedAirRequestIds.push(selectedAirRequestId);
      dispatch({
        type: SET_SELECTED_AIR_REQUEST_ID,
        payload: {
          selectedAirRequestIds
        }
      });
    }
  };
}

/**
 * To remove value from selectedAirRequestIds in state
 * @param {Arrar<string>} idsToRemove
 */
export function removeSelectedAirRequestIds(idsToRemove = []) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedAirRequestIds = getState().travelCart.userSession.airRequest.selectedAirRequestIds || [];

    if (selectedAirRequestIds.length > 0 && idsToRemove.length > 0) {
      const filteredIds = selectedAirRequestIds.filter(existingId => {
        if (!idsToRemove.includes(existingId)) {
          return existingId;
        }
      });
      dispatch({
        type: SET_SELECTED_AIR_REQUEST_ID,
        payload: {
          selectedAirRequestIds: filteredIds
        }
      });
    }
  };
}

/**
 * To add value to selectedAirActualIds in state
 */
export function setSelectedAirActualId(selectedAirActualId: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedAirActualIds = [...getState().travelCart.userSession.airActual.selectedAirActualIds];
    if (!selectedAirActualIds.includes(selectedAirActualId)) {
      selectedAirActualIds.push(selectedAirActualId);
      dispatch({
        type: SET_SELECTED_AIR_ACTUAL_ID,
        payload: {
          selectedAirActualIds
        }
      });
    }
  };
}

/**
 * To remove value from selectedAirActualIds in state
 * @param {Arrar<string>} idsToRemove
 */
export function removeSelectedAirActualIds(idsToRemove = []) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedAirActualIds = getState().travelCart.userSession.airActual.selectedAirActualIds || [];

    if (selectedAirActualIds.length > 0 && idsToRemove.length > 0) {
      const filteredIds = selectedAirActualIds.filter(existingId => {
        if (!idsToRemove.includes(existingId)) {
          return existingId;
        }
      });
      dispatch({
        type: SET_SELECTED_AIR_ACTUAL_ID,
        payload: {
          selectedAirActualIds: filteredIds
        }
      });
    }
  };
}
