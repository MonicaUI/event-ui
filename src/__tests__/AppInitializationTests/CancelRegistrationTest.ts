import renderMainApp from '../../testUtils/renderMainApp';
import * as eventStatus from 'event-widgets/clients/EventStatus';
import EventGuestClient from '../../clients/EventGuestClient';
import RegCartClient from '../../clients/RegCartClient';

const mockEventId = '11111111-2222-3333-4444-555555555555';
const eventId = mockEventId;
const fakePath = `/${eventId}/`;
const wait = ms => new Promise(resolve => setTimeout(resolve, ms));

import { getRegCart } from '../../redux/selectors/shared';
import { restoreTravelCartIntoState, startTravelCancellation } from '../../redux/travelCart/external';

jest.mock('../../redux/travelCart/external', () => {
  return {
    ...jest.requireActual<$TSFixMe>('../../redux/travelCart/external'),
    __esModule: true,
    restoreTravelCartIntoState: jest.fn(),
    startTravelCancellation: jest.fn()
  };
});

jest.mock('../../redux/selectors/shared', () => {
  return {
    ...jest.requireActual<$TSFixMe>('../../redux/selectors/shared'),
    __esModule: true,
    getRegCart: jest.fn()
  };
});

(restoreTravelCartIntoState as $TSFixMe).mockImplementation(() => {
  return dispatch => {
    dispatch({
      type: '[MOCK]/restoreTravelCartIntoState',
      payload: {}
    });
  };
});
(startTravelCancellation as $TSFixMe).mockImplementation(() => {
  return dispatch => {
    dispatch({
      type: '[MOCK]/startTravelCancellation',
      payload: {}
    });
  };
});

const mockConfirmedRegCartResponse = {
  regCart: {
    regCartId: 'regCartId',
    status: 'INPROGRESS',
    regCancel: true,
    eventSnapshotVersions: {
      [mockEventId]: 'eventSnapshotVersion'
    },
    eventRegistrations: {
      'event-registrations-key': {
        confirmationNumber: 'fake-number',
        attendee: {
          attendeeId: 'fake-attendeeId',
          personalInformation: {
            emailAddress: 'fake-email-address'
          }
        },
        registrationPathId: 'todoThisShouldBeRegPathId',
        sessionWaitlists: {}
      }
    }
  },
  validationMessages: []
};

jest.mock('../../clients/RegCartClient');
(RegCartClient.prototype.getRegCart as $TSFixMe).mockImplementation(() =>
  Promise.resolve(mockConfirmedRegCartResponse.regCart)
);
(RegCartClient.prototype.getCapacitySummaries as $TSFixMe).mockImplementation(() => Promise.resolve({}));
(RegCartClient.prototype.identifyByConfirm as $TSFixMe).mockImplementation(() =>
  Promise.resolve(mockConfirmedRegCartResponse)
);
(RegCartClient.prototype.createCancelRegistrationCart as $TSFixMe).mockImplementation(() =>
  Promise.resolve(mockConfirmedRegCartResponse)
);

jest.mock('../../clients/EventGuestClient');
EventGuestClient.prototype.publishAttendeeActivityFact = jest.fn(() => Promise.resolve({}));

test('cancelRegistration starts a cancel regCart and goes to cancellation page', async () => {
  (getRegCart as $TSFixMe).mockImplementation(() => mockConfirmedRegCartResponse.regCart);
  const mainApp = await renderMainApp(
    {
      eventContext: {
        eventId,
        regCartId: 'fake-reg-cart-id',
        confirmationNumber: 'fake-confirmation-number'
      },
      eventLaunchWizardSettings: '{}',
      accessToken: 'BEARER fakeToken'
    },
    fakePath + 'cancelRegistration',
    event => ({
      ...event,
      id: mockEventId,
      status: eventStatus.ACTIVE
    })
  );
  await wait(0);
  expect(mainApp.getDOMNode()).toMatchSnapshot();
  expect(RegCartClient.prototype.createCancelRegistrationCart).toHaveBeenCalledWith(
    'BEARER fakeToken',
    undefined,
    undefined,
    mockEventId,
    null
  );
  expect(RegCartClient.prototype.identifyByConfirm).not.toHaveBeenCalled();

  // verify that reg cancel flow also initiates travel cancellation
  expect(startTravelCancellation).toBeCalledWith('regCartId');
});

test('planner cancelRegistration logs in registrant with plannerRegSettings', async () => {
  jest.clearAllMocks();
  (getRegCart as $TSFixMe).mockImplementation(() => mockConfirmedRegCartResponse.regCart);
  const plannerViewRoot = `/planner-path/${eventId}`;
  const mainApp = await renderMainApp(
    {
      eventContext: {
        eventId,
        isPlanner: true
      },
      eventLaunchWizardSettings: '{}',
      accessToken: 'BEARER plannerRegModToken-from-html',
      viewRoot: plannerViewRoot, // load planner site on same fake root just for this test
      plannerRegSettings: {
        successUrl: '/fakeUrl',
        exitUrl: '/fakeUrl',
        modificationRequest: {
          emailAddress: 'abc@123.com',
          confirmationNumber: 'abc123'
        }
      }
    },
    `${plannerViewRoot}/cancelRegistration`,
    event => ({
      ...event,
      id: mockEventId,
      status: eventStatus.ACTIVE
    })
  );
  await wait(0);
  expect(mainApp.getDOMNode()).toMatchSnapshot();
  expect(RegCartClient.prototype.createCancelRegistrationCart).toHaveBeenCalledWith(
    'BEARER plannerRegModToken-from-html',
    'fake-attendeeId',
    'fake-number',
    mockEventId,
    null
  );
  expect(RegCartClient.prototype.identifyByConfirm).toHaveBeenCalled();
  expect(restoreTravelCartIntoState).not.toHaveBeenCalled();
});
