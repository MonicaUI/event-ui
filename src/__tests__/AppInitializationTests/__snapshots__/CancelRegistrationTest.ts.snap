// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cancelRegistration starts a cancel regCart and goes to cancellation page 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="container"
      style="background-color: rgb(4, 21, 50); border-radius: 0;"
    >
      <div
        style="max-width: 1024px; margin: auto;"
      >
        <div>
          <div>
            <div>
              <div>
                <div
                  role="banner"
                >
                  <div
                    class="mockReadOnlyContent"
                  >
                    container:7162204f-7d0c-4be9-ad13-3fe7dec332bd
                  </div>
                </div>
                <div
                  id="main"
                  role="main"
                  tabindex="-1"
                >
                  <div
                    class="mockReadOnlyContent"
                  >
                    container:beddb7b5-5e71-4103-8ab4-a9cbc9078808
                  </div>
                </div>
                <div
                  class="mockReadOnlyContent"
                >
                  container:8471fca9-0b98-41df-81a4-a2b5944629a1
                </div>
              </div>
              <div
                role="contentinfo"
              >
                <div
                  class="mockReadOnlyContent"
                >
                  CventFooterContainer
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</div>
`;

exports[`planner cancelRegistration logs in registrant with plannerRegSettings 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="container"
      style="background-color: rgb(4, 21, 50); border-radius: 0;"
    >
      <div
        style="max-width: 1024px; margin: auto;"
      >
        <div>
          <div>
            <div
              class="banner"
              data-cvent-id="planner-reg-page-banner"
            >
              <span>
                EventGuestSide_BannerText_PlannerReg__resx
              </span>
              <div
                class="bannerHelp"
              >
                <span>
                  EventGuestSide_BannerHelpText_PlannerReg__resx
                </span>
              </div>
              <div
                class="actionSection"
              >
                <div
                  class="aside"
                />
                <ul
                  class="actionList"
                >
                  <li
                    class="actionExit"
                  >
                    <button
                      aria-label="_close__resx"
                      class="exitEditor"
                      data-cvent-id="page-banner-close-button"
                      type="button"
                    >
                      <span
                        class="container"
                      >
                        <span
                          aria-label="_close__resx"
                          class="exitIcon"
                          role="img"
                        />
                      </span>
                    </button>
                  </li>
                </ul>
              </div>
            </div>
            <div>
              <div>
                <div
                  role="banner"
                >
                  <div
                    class="mockReadOnlyContent"
                  >
                    container:7162204f-7d0c-4be9-ad13-3fe7dec332bd
                  </div>
                </div>
                <div
                  id="main"
                  role="main"
                  tabindex="-1"
                >
                  <div
                    class="mockReadOnlyContent"
                  >
                    container:beddb7b5-5e71-4103-8ab4-a9cbc9078808
                  </div>
                </div>
                <div
                  class="mockReadOnlyContent"
                >
                  container:8471fca9-0b98-41df-81a4-a2b5944629a1
                </div>
              </div>
              <div
                role="contentinfo"
              >
                <div
                  class="mockReadOnlyContent"
                >
                  CventFooterContainer
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</div>
`;
